In [ ]:
import numpy as np 
import pandas as pd 

%load_ext autoreload
%autoreload 2
import funciones_etl
import warnings
warnings.filterwarnings("ignore")

### Se extraen los datos, pestaña HECHOS

In [17]:
# Abrimos el archivo excel
homicidios_hechos = pd.read_excel("../Datos/homicidios.xlsx", sheet_name="HECHOS")
homicidios_hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

Entre los datos generales, una vez abierto este archivo podemos encontrar que tiene 21 columnas y 696 filas

In [18]:
# Observar columnas
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

In [19]:
# Observar cantidad de registros o filas
homicidios_hechos.shape[0]

696

Se puede corroborar la cantidad de columnas y de registros en este dataframe

### Normalizar encabezados de las columnas

Como algunas columnas tienen su nombre en mayusaculas otras en minusculas ademas de que unas cuentan con espacios y otras con "_" en su lugas decidimos estandarizar dichas columnas, de manera que eliminiaremos los esapacios pensando en cargar la data en una base de datos con motor mysql

In [22]:
# Capitalize, para colocar la primera letra en mayuscula
homicidios_hechos.columns = [x.capitalize() for x in homicidios_hechos.columns]

#Se reemplazan los espacios por guion bajo "-"
homicidios_hechos.columns = homicidios_hechos.columns.str.replace(" ", "_")

# Se renombran algunas columnas para darle mas sentido al dataframe
homicidios_hechos = homicidios_hechos.rename(columns={"Id": "Id_hecho",
                                                      "N_victimas": "Cantidad_victimas",
                                                      "Aaaa": "Año",
                                                      "Mm": "Mes",
                                                      "Dd": "Dia",
                                                      "Hh": "Hora_entero"})

In [24]:
homicidios_hechos.columns

Index(['Id_hecho', 'Cantidad_victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Hora_entero', 'Lugar_del_hecho', 'Tipo_de_calle', 'Calle', 'Altura',
       'Cruce', 'Dirección_normalizada', 'Comuna', 'Xy_(caba)', 'Pos_x',
       'Pos_y', 'Participantes', 'Victima', 'Acusado'],
      dtype='object')

### Valores nulos

In [29]:
nulos_tipo = funciones_etl.verificar_datos(homicidios_hechos)
nulos_tipo

Advertencia: la columna Calle, tiene valores nulos
Advertencia: la columna Altura, tiene valores nulos
Advertencia: la columna Cruce, tiene valores nulos
Advertencia: la columna Dirección_normalizada, tiene valores nulos


columna                                          tipo_dato  \
0                Id_hecho                                    [<class 'str'>]   
1       Cantidad_victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7             Hora_entero                     [<class 'int'>, <class 'str'>]   
8         Lugar_del_hecho                                    [<class 'str'>]   
9           Tipo_de_calle                                    [<class 'str'>]   
10                  Calle                   [<class 'str'>, <class 'float'>]   
11                 Altura                                  [<class 'float'>]   
12                  Cruce                   [<class 'str'>, <class 'float'>]   
13  Dirección_normalizada                   [<class 'str'>, <class 'float'>]   
14                 Comuna                                    [<class 'int'>]   
15              Xy_(caba)                                    [<class 'str'>]   
16                  Pos_x                                    [<class 'str'>]   
17                  Pos_y                                    [<class 'str'>]   
18          Participantes                                    [<class 'str'>]   
19                Victima                                    [<class 'str'>]   
20                Acusado                                    [<class 'str'>]   

    datos_nulos  porcentaje_nulos  porcentaje_no_nulos  
0             0              0.00               100.00  
1             0              0.00               100.00  
2             0              0.00               100.00  
3             0              0.00               100.00  
4             0              0.00               100.00  
5             0              0.00               100.00  
6             0              0.00               100.00  
7             0              0.00               100.00  
8             0              0.00               100.00  
9             0              0.00               100.00  
10            1              0.14                99.86  
11          567             81.47                18.53  
12          171             24.57                75.43  
13            8              1.15                98.85  
14            0              0.00               100.00  
15            0              0.00               100.00  
16            0              0.00               100.00  
17            0              0.00               100.00  
18            0              0.00               100.00  
19            0              0.00               100.00  
20            0              0.00               100.00

Podemos observar 747 datos nulos en el dataframe, repartidos de la siguiente manera

* Columna Altura = 567 datos nulos, que representan un 81.47 % de sus datos

* Columna Cruce = 171 datos nulos, que representan 24.57 % de sus datos

* Columna Calle = 1 datos nulos, que representan el 0.14 % de sus datos

* Columna Direccion_normalizada = 8 datos nulos, que representan 1,15 % de sus datos

Altura: 

Comparamos los vaores de las columnas Altura, Lugar_del_hecho y Direcccion_normalizada, esto debido a que las res columnas se trata de datos sobre la direccion donde ocurre el siniestro asi que de esta manera buscamos coincidencias que nos permitan completar informacion

In [33]:
# Comparamos los valores de las columnas
df_no_nulos_altura = homicidios_hechos[["Altura", "Lugar_del_hecho", "Dirección_normalizada"]]
df_no_nulos_altura

Altura                            Lugar_del_hecho  \
0       NaN  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ   
1       NaN          AV GRAL PAZ  Y AV DE LOS CORRALES   
2    2034.0                         AV ENTRE RIOS 2034   
3       NaN      AV LARRAZABAL Y GRAL VILLEGAS CONRADO   
4       NaN  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA    
..      ...                                        ...   
691     NaN                          AV. RIESTRA Y MOM   
692     NaN                    AU DELLEPIANE Y LACARRA   
693     NaN                        AV. GAONA Y TERRADA   
694  4071.0                         AV. EVA PERON 4071   
695   709.0                    PADRE CARLOS MUJICA 709   

                                 Dirección_normalizada  
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.  
2                                  ENTRE RIOS AV. 2034  
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.  
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.  
..                                                 ...  
691                                  RIESTRA AV. y MOM  
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.  
693                                GAONA AV. y TERRADA  
694                                PERON, EVA AV. 4071  
695                            PADRE CARLOS MUGICA 709  

[696 rows x 3 columns]

Obtenemos los datos no nulos y los mostramos

In [34]:
no_nulos_altura = df_no_nulos_altura.loc[~df_no_nulos_altura.isnull().any(axis=1)]

In [35]:
no_nulos_altura

Altura          Lugar_del_hecho    Dirección_normalizada
2    2034.0       AV ENTRE RIOS 2034      ENTRE RIOS AV. 2034
9    1366.0       AV ENTRE RIOS 1366      ENTRE RIOS AV. 1366
14    156.0             SUIPACHA 156             SUIPACHA 156
33   1483.0                LIMA 1483                LIMA 1483
42     30.0  AV LUIS MARIA CAMPOS 30   CAMPOS, LUIS M. AV. 30
..      ...                      ...                      ...
677  3112.0             TUCUMAN 3112             TUCUMAN 3112
685  1359.0           ESMERALDA 1359           ESMERALDA 1359
687  3655.0           AV. GAONA 3655           GAONA AV. 3655
694  4071.0       AV. EVA PERON 4071      PERON, EVA AV. 4071
695   709.0  PADRE CARLOS MUJICA 709  PADRE CARLOS MUGICA 709

[129 rows x 3 columns]

Podemos observar que los datos que no son nulos de la columnas Altura se repiten en las columnas Lugar_delhecho y Direccion_normalizada, por lo tanto se decide eliminar la columna

In [36]:
homicidios_hechos = homicidios_hechos.drop("Altura", axis=1)
homicidios_hechos.columns

Index(['Id_hecho', 'Cantidad_victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Hora_entero', 'Lugar_del_hecho', 'Tipo_de_calle', 'Calle', 'Cruce',
       'Dirección_normalizada', 'Comuna', 'Xy_(caba)', 'Pos_x', 'Pos_y',
       'Participantes', 'Victima', 'Acusado'],
      dtype='object')